### PR TITLE
feat: added support for new custom function parameter data types

### PR DESCRIPTION
**Pull request by abhilash-kumar-nair**
Jira - https://modelonproducts.atlassian.net/browse/FLOW-483
This PR adds support for CaseResult and ExperimentResult datatypes in python client

### CODE REVIEW

**Review by iakovn — COMMENTED** [review]
Nice start!
I'm wondering if we need to reconsider update to:
```
class _Parameter:
 @property
    def value(self) -> Any:
        return self._value
class CustomFunction:
     def parameter_values(self) -> Dict[str, Any]:
```
It seems right now if you set those for ExperimentResult/CaseResult type you need to use Experiment/Case object but when reading back it returns just strings. 

Do you have some examples in mind on how it will work when recreating from previously run experiment/case?

**Comment by abhilash-kumar-nair** [comment]
> Nice start! I'm wondering if we need to reconsider update to:
> 
> ```
> class _Parameter:
>  @property
>     def value(self) -> Any:
>         return self._value
> class CustomFunction:
>      def parameter_values(self) -> Dict[str, Any]:
> ```
> 
> It seems right now if you set those for ExperimentResult/CaseResult type you need to use Experiment/Case object but when reading back it returns just strings.
> 
> Do you have some examples in mind on how it will work when recreating from previously run experiment/case?

Good point! I have created a new class that inherits the built in dict with an additional method that does the conversion to raw dict(that is to be sent to the back-end). Does this make sense?

**Review by iakovn — CHANGES_REQUESTED** [review]
This is fine in one direction. I think we need the reverse as well. 🔙 

Consider:
```
definition = case.get_definition()
cf = definition.custom_function
init_from_case = cf.initialize_from_case
pv = cf.parameter_values
baseline_case = pv['baseline']
```
I expect both `init_from_case` and  `baseline_case` to be either `Case` entity or some new `CaseReference` objects (or None).

I would consider naming to/from protocol somewhere.

File: modelon/impact/client/entities/case.py [hunk 1]
```diff
@@ -789,16 +790,44 @@ def execute(self, sync_case_changes: bool = True) -> CaseOperation:
...
+        custom_function_params_override: Dict[str, Any],
+    ) -> None:
+        for param in custom_function._param_by_name.values():
+            param_override = custom_function_params_override.get(param.name)
+            if param_override:
+                if param.type == "ExperimentResult":
+                    exp_info = self._sal.workspace.experiment_get(
+                        self._workspace_id, param_override
+                    )
+                    custom_function_params_override[param.name] = ExperimentReference(
```
[the remark this answers is not in the document]
Comment (abhilash-kumar-nair) [reply modelon/impact/client/entities/case.py]: Yes it was the circular dependencies. I have now updated the code to return references instead and a utility method to create entities from references.

File: modelon/impact/client/client.py [hunk 2]
```diff
@@ -741,6 +741,22 @@ def get_me(self) -> User:
     def create_entity_from_reference(
```
Comment (iakovn) [inline modelon/impact/client/client.py:741]: I suggest for user facing flow two functions marked "experimental":
```
def get_experiment_by_reference(...) -> Experiment
def get_case_by_reference(...) -> Case
```

**Review by iakovn — CHANGES_REQUESTED** [review]
Let's make two experimental functions and merge. I'd like to play in production before finalizing

File: modelon/impact/client/client.py (hunk 2 above)
Comment (abhilash-kumar-nair) [reply modelon/impact/client/client.py]: Makes sense! Done!

**Review by iakovn — APPROVED** [review]
Let's merge! ✈️

**Pull request merged**